lib/Dialect/Arc/ArcCostModel.cpp

```diff
@@ -0,0 +1,106 @@
+//===- ArcCostModel.cpp ---------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Arc/ArcCostModel.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include <algorithm>
+
+using namespace llvm;
+using namespace circt;
+using namespace arc;
+using namespace std;
+
+size_t ArcCostModel::getCost(Operation *op) { return computeOperationCost(op); }
+
+size_t ArcCostModel::computeOperationCost(Operation *op) {
+  if (isa<circt::comb::ConcatOp>(op))
+    return size_t(OperationCost::CONCATCOST);
+  if (isa<circt::comb::ExtractOp>(op))
+    return size_t(OperationCost::EXTRACTCOST);
+  // We have some other functions that need to be handled in a different way
+  // arc::StateOp, arc::CallOp, mlir::func::CallOp and arc::VectorizeOp, each of
+  // these functions have bodies so the cost of the op equals the cost of its
+  // body.
+  if (isa<arc::StateOp>(op) || isa<arc::CallOp>(op) ||
+      isa<mlir::func::CallOp>(op))
+    return countBodyOps(op->getOperand(0).getDefiningOp()->getRegions());
```

Comment:
`getOperand(0)` gives you the first call argument. I think what you want here is `getCallee()` or similar. I think the `CallOpInterface` that is implemented by these ops might have a few helpful functions :smiley: